group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -428,9 +431,32 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        CompletableFuture<ListGroupsResponseData> future = new CompletableFuture<>();
+        List<ListGroupsResponseData.ListedGroup> results = new ArrayList<>();
+        final AtomicInteger cnt = new AtomicInteger(runtime.partitions().size());
+
+        for (TopicPartition tp : runtime.partitions()) {
```

Comment:
It seems to me that calling `partitions()` twice is not safe here because the number of partitions could change in between the two calls. I think that we should store it in order to avoid this race condition.